posts/2018-05-09-tagless-final-streaming.md

```diff
@@ -0,0 +1,172 @@
+  override def save(item: Item): Id[Unit] = mem.update(item.name.value, item)
+  override def remove(name: ItemName): Id[Unit] = {
+    mem.remove(name.value)
+    ()
+  }
+}
+```
+
+That's pretty much it! We managed to abstract over the return type of `findAll` by just adding an extra parameter to our algebra.
+
+### Source of inspiration
+
+I've come up with most of the ideas presented here during my work on [Fs2 Rabbit](https://gvolpe.github.io/fs2-rabbit/), a stream based client for `Rabbit MQ`, where I make heavy use of this technique as I originally described in [this blog post](https://partialflow.wordpress.com/2018/02/01/a-tale-of-tagless-final-cats-effect-and-streaming-fs2-rabbit-v0-1/).
+
+Another great source of inspiration was [this talk](https://www.youtube.com/watch?v=1h11efA4k8E) given by [Luka Jacobowitz](https://github.com/LukaJCB) at Scale by the Bay.
+
+### Abstracting over the effect type
+
+One thing you might have noticed in the examples above is that both `ItemRepository` interpreters are not fixed to `IO` or `Task` or any other effect type but rather requiring a parametric `F[_]` and an implicit instance of `Sync[F]`. This is a quite powerful technique for both library authors and application developers. Well know libraries such as [Http4s](https://http4s.org/), [Monix](https://monix.io/) and [Fs2](https://functional-streams-for-scala.github.io/fs2/) make a heavy use of it.
+
+And by requiring a `Sync[F]` instance we are just saying that our implementation will need to suspend synchronous side effects.
+
+Once at the edge of our program, commonly the main method, we can give `F[_]` a concrete type. At the moment, there are two options: `cats.effect.IO` and `monix.eval.Task`. But hopefully soon we'll have a `Scalaz 8 IO` implementation as well (fingers crossed).
+
+### Principle of least power
+
+Abstracting over the effect type doesn't only mean that we should require `Sync[F]`, `Async[F]` or `Effect[F]`. It also means that we should only require the minimal typeclass instance that satisfies our predicate. For example:
+
+```scala
+import cats.Functor
+import cats.implicits._
+
+def bar[F[_]: Applicative]: F[Int] = 1.pure[F]
+
+def foo[F[_]: Functor](fa: F[Int]): F[String] =
+  fa.map(_.toString)
+```
+
+Here our `bar` method just returns a pure value in the `F` context, thus we need an `Applicative[F]` instance that defines `pure`. On the other hand, our `foo` method just converts the inner `Int` into `String`, what we call a pure data transformation. So all we need here is a `Functor[F]` instance. Another example:
+
+```scala
+import cats.Monad
+
+def fp[F[_]: Monad]: F[String] =
+`  for {
+    a <- bar[F]
+    b <- bar[F]
+  } yield a + b
+```
+
+The above implementation makes use of a `for-comprehention` which is a syntactic sugar for `flatMap` and `map`, so all we need is a `Monad[F]` instance because we also need an `Applicative[F]` instance for `bar`, otherwise we could just use a `FlatMap[F]` instance.
```

Comment:
This is also a so-so example. `(bar[F], bar[F]).mapN(_ + _)` would only require `Applicative` and because of consistency law would also work for any `Monad` in exactly the same way.

Reply:
The idea here was to show that by using a for-comprehention `flatMap` is required, hence one will need either a `FlatMap` or `Monad` constraint. I understand that with `mapN` only an `Applicative` is needed but the idea here is a bit different, not sure if I get your point?